index.js

```diff
@@ -448,20 +448,22 @@ class TuyaDevice extends EventEmitter {
     }
 
     /**
-     * Emitted when data is returned from device.
+     * Emitted when data is returned from device unless it's a bad response to DP_QUERY
      * @event TuyaDevice#data
      * @property {Object} data received data
      * @property {Number} commandByte
      * commandByte of result
      * (e.g. 7=requested response, 8=proactive update from device)
      * @property {Number} sequenceN the packet sequence number
      */
-    this.emit('data', packet.payload, packet.commandByte, packet.sequenceN);
+    if (!(packet.commandByte === CommandType.DP_QUERY &&
+          packet.payload === 'json obj data unvalid')) {
```

Comment:
There's a bunch of other possible error strings, so do we check for all of them here?

I feel like `data` should always be emitted regardless of what the received data actually means. At this point it's a design limitation; but since this is gonna be a breaking change anyways what do you think about adding another event (`state` or similar) that *only* emits when the device returns a valid JSON object (in addition to the current `data` event)?

Reply:
It's a valid point, but the reason I'm suppressing this one is because it's being caused by the API itself yet it's being surfaced to the API consumer.  As an API consumer I call get(), which ends up causing a data event sent to my consuming application.  So as that consumer what can I do with it?  It's not useful to me and my code didn't cause it (I just issued a get) so my upper level code has to suppress/ignore it.  Personally, I think it only makes sense to surface data from the device that is unsolicited (i.e. data update events) because everything else is useless to me as an API consumer anyway, so it's just noise.  Based on this I'd argue suppress all DP_QUERY responses since, as far as I know, those never come unsolicited from the device so they have to be a response to the API, not to the API consumer.

Reply:
> As an API consumer I call get(), which ends up causing a data event sent to my consuming application. So as that consumer what can I do with it? It's not useful to me and my code didn't cause it (I just issued a get) so my upper level code has to suppress/ignore it.

Absolutely, but some applications may want errors to be bubbled up. And for some errors (like when the timestamp field is wrong), it starts to get a little muddy on whether it's the package's fault or the consumer's fault.

> Based on this I'd argue suppress all DP_QUERY responses since, as far as I know, those never come unsolicited from the device so they have to be a response to the API, not to the API consumer.

I think you mean suppress everything **except** DP_QUERY responses, unless I misinterpreted the rest?

In any case, we can take this somewhere else. Probably outside the scope of this PR.